src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs

```diff
@@ -15,11 +15,86 @@ namespace Microsoft.AspNetCore.Hosting
     public static class WebHostExtensions
     {
         /// <summary>
+        /// Blocks the calling thread until the web application shuts down.
+        /// </summary>
+        /// <param name="host">The <see cref="IWebHost"/> to wait for shutdown</param>
+        public static void WaitForShutdown(this IWebHost host)
+        {
+            WaitForSystemShutdown(host, (token, message) => host.WaitForShutdown(token, message));
+        }
+
+        /// <summary>
+        /// Blocks the calling thread until the web application shuts down.
+        /// </summary>
+        /// <param name="host">The <see cref="IWebHost"/> to wait for shutdown</param>
+        /// <param name="token">The token to trigger shutdown.</param>
+        public static void WaitForShutdown(this IWebHost host, CancellationToken token)
+        {
+            host.WaitForShutdown(token, shutdownMessage: null);
+        }
+
+        /// <summary>
         /// Runs a web application and block the calling thread until host shutdown.
         /// </summary>
         /// <param name="host">The <see cref="IWebHost"/> to run.</param>
         public static void Run(this IWebHost host)
         {
+            WaitForSystemShutdown(host, (token, message) => host.Run(token, message));
+        }
+
+        /// <summary>
+        /// Runs a web application and block the calling thread until token is triggered or shutdown is triggered.
+        /// </summary>
+        /// <param name="host">The <see cref="IWebHost"/> to run.</param>
+        /// <param name="token">The token to trigger shutdown.</param>
+        public static void Run(this IWebHost host, CancellationToken token)
+        {
+            host.Run(token, shutdownMessage: null);
+        }
+
+        private static void Run(this IWebHost host, CancellationToken token, string shutdownMessage)
+        {
+            using (host)
+            {
+                host.Start();
+
+                host.WaitForShutdown(token, shutdownMessage);
+            }
+        }
+
+        private static void WaitForShutdown(this IWebHost host, CancellationToken token, string shutdownMessage)
+        {
+            var hostingEnvironment = host.Services.GetService<IHostingEnvironment>();
+            var applicationLifetime = host.Services.GetService<IApplicationLifetime>();
+
+            Console.WriteLine($"Hosting environment: {hostingEnvironment.EnvironmentName}");
```

Comment:
There's a whole discussion on this too. I think the default should use the console. Having an options to turn it off would be fine though.